### DISCUSSION TITLE
HyperSPI with LolinS2-Mini Pin Output in Multi Segment

### DISCUSSION THREAD

**Discussion by skalpjaeger** [awawa-dev/HyperSPI#24] [General]
Hi

I wanna change the LED Pinout from the LolinS2-Mini 2 to 5 ist that possible (I know I have to edit the platform.ini)? And which is then the correct GPIO for the RP4?

Thank you very much

**Comment by awawa-dev** [comment]
Hi
I think GPIO5 is not reserved on ESP32-S2 mini so you can use it but better test it before soldering anything (I didnt test it). The pinout for the ESP32-S2 is provided on the main page and the Rpi4 GPIO schematic/list can be easily found online so is there anything unclear?

**Comment by skalpjaeger** [comment]
Thank you for the fast reply. I am (sorry for that), however a little bit confused about the multiple pictures. I have soldered the PCB Board exactly like you advised to to. I have GPIO 16 and 33 on the S2 for the LED output on the S2. So I have to declare 16 and 33 in platform.io for the main and second segment as data pins right? I also know that I have to connect GPIO 7, 11, 12+ ground from the S2 mini with the raspberry pi. However, in your picture of your instruction (https://user-images.githubusercontent.com/69086569/218791687-0d1eb9ab-e28c-4997-80d5-c2f3b5e3ee3c.jpg ) there are 4 wires+ground to the raspi and this is what confuses me.... I know its with the esp32 mh-et live but did I miss something?

Thank you very much for your help

**Reply by awawa-dev** [reply]
On this picture there is a full SPI connection but SPI:MISO cable is not needed by HyperSPI that's why only 3 SPI cables are used + ground (it's very important). In the latest image on the HyperSPI main page, the miso cable is no longer used:
![obraz](https://user-images.githubusercontent.com/69086569/231194107-33ae9d80-a1c7-40c0-aad4-94a45ef59f9a.png)

**Reply by skalpjaeger** [reply]
Ok perfekt thank you very much!!

**Comment by awawa-dev** [comment]
![obraz](https://user-images.githubusercontent.com/69086569/231207350-a670bfea-a96d-4d21-9e8f-f2ca027105da.png)

**Comment by skalpjaeger** [comment]
Is there a way to manually start the workflow? because I forked your repository, edited platform.io but artefact doesn't appear in actions... I have granted the permission like  you said in your description.

**Reply by awawa-dev** [reply]
Exactly as it says in the manual: you must enable Github Action **before** you make any changes to the repository to run the action. If you enabled it later, just edit any file (even readme.md) and commit changes again. This will trigger the build of the whole firmware package.

**Comment by skalpjaeger** [comment]
Ok weird I did it exactly like this, but now I edited the push-master.yml and then it worked...Thanks!

**Comment by skalpjaeger** [comment]
Hey Awawa

Sorry for bothering again, but is there a way to see in hyperhdr to see if the esp32 s2 is properly connected to the raspi? Because unfortunately my leds are not working and im looking for the error...

Thanks again for your help!

**Reply by awawa-dev** [reply]
Yes, https://github.com/awawa-dev/HyperSPI/discussions/17 and in particular https://github.com/awawa-dev/HyperSPI/discussions/10#discussioncomment-4366604 . Logs to the serial output are provided only if there is no communication with HyperHDR so HyperHDR service must be stopped or HyperHDR HyperSPI instance LED must be stopped.

**Reply by awawa-dev** [reply]
Do you have any USB serial output logs at all of the ESP32-S2 mini if you are using screen under Linux or PuTTY under Windows (without SPI connection to the Rpi)?

**Reply by skalpjaeger** [reply]
I couldn't install screen since I have the raspi running on libreelec...
I think I am very lucky to got 3 broken lolin s2... I tried the hole thing with HyperSerialESP32 and the first 2 boards weren't even recognised by the raspi, the third one gets recognised but the leds are still not working although im pretty sure that the connection is correct.

**Reply by awawa-dev** [reply]
While it's possible, I highly doubt that all three are broken. Possibly came clean without bootloader or bootloader is corrupted which may cause the board not to be detected by the system (esp32-s2 serial port is virtual, it doesn't work without loaded firmware). I suggest you start with HyperSerialESP32 as it's easier to debug (you can also use HyperHDR on Windows to run and test this), then you can switch to HyperSPI on Rpi if you want. Follow the esp32-s2 mini recovery procedure described here: https://github.com/awawa-dev/HyperSerialESP32#flashing

**Comment by skalpjaeger** [comment]
yeah I already did the recovery procedure a few times, like I said the actual lolin mini is recognised by hyperhdr but leds weren't working. However when checking my order I saw that I ordered the wrong led strip, actually I wanted the cold one but ordered the neutral warm, so I always flashed the wrong firmware. So I hope that was my mistake and the leds will turn on with the correct one. if not I think that I have to order a new esp...

**Reply by awawa-dev** [reply]
The cold and neutral version of sk6812 is handled in a similar way, the neutral needs different color calibration before sending to the led strip. But the communication protocol is identical for both versions. Verify that there is communication between HyperHDR and HyperSerialESP32: https://github.com/awawa-dev/HyperSerialESP32#usage-in-hyperhdr (testing performance/reading statistics). If the communication is correct, it is probably an electrical problem.

**Reply by awawa-dev** [reply]
https://www.hyperhdr.eu/2023/02/ultimate-guide-on-how-to-build-led.html#chapter9

**Reply by skalpjaeger** [reply]
although the lolin s2 is detected and selected in the adalight driver, there is no communication, frames are 0 and at the overview page the LED-Device is pending all the time (Which was not the case at the beginning (once the led device was loaded, frames were around 60) when I used HyperSPI, but im pretty sure that these eps weren't working correctly)...

I did the recovery procedure for 2 times now (facory.bin) and I do not know what else I can do. I use libreelec but that shouldn't be a problem or? Do you have any further ideas? maybe really three broken eps?

**Reply by awawa-dev** [reply]
Please upload your exported HyperHDR configuration and attach HyperHDR logs (that part related to open and close of the LED device). If you are reading statistics (even if there is 0fps) that means your esp is working correctly. Probably some configuration issues or maybe you didn't enable continue output in the smoothing when you tested the color effect

**Reply by awawa-dev** [reply]
Are you using v19 version?

**Reply by skalpjaeger** [reply]
Continue output was one solution, the other was to change the baud rate to 3000000 since it did not work with 5000000. Do you have an explanation for this?

**Reply by awawa-dev** [reply]
No, in the case of esp32-s2 lolin CDC it's irrelevant. What is important is that it is definitely a continuous output during testing. But it was mentioned in every guide.

**Reply by skalpjaeger** [reply]
Im really sorry for missing that! the other thing was possibly a missing ground. It was working yesterday with a breadboard and in another try to use my pcb I damaged the lolin (Its not regognised by the raspi nor my pc again). I have one left which is regognised by my Mac and by the raspi (and flashing firmware is also possible) but LEDs are dark now. In the logs there is the following 

> 2023-04-16T15:37:10.690Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:216) _rs232Port.open(QIODevice::ReadWrite): ttyACM0, Baud rate [5000000]bps
2023-04-16T15:37:10.727Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:221) portName:          ttyACM0
2023-04-16T15:37:10.728Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:222) systemLocation:    /dev/ttyACM0
2023-04-16T15:37:10.728Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:223) description:       LOLIN-S2-MINI
2023-04-16T15:37:10.728Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:224) manufacturer:      WEMOS.CC
2023-04-16T15:37:10.728Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:225) productIdentifier: 0x80c2
2023-04-16T15:37:10.728Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:226) vendorIdentifier:  0x303a
2023-04-16T15:37:10.728Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:227) serialNumber:      0
2023-04-16T15:37:10.730Z [LEDDEVICE_ADALIGHT] Detected ESP32-S2 lolin mini type board. HyperHDR skips the reset. State: 1, 1
2023-04-16T15:37:10.734Z [LEDDEVICE_ADALIGHT] ESP sent: 'HyperHDR frames: 0 (FPS), receiv.: 0, good: 0, incompl.: 0, mem1: 3508, mem2: 3720, heap: 133464'
2023-04-16T15:37:10.735Z [LEDDEVICE_ADALIGHT] ESP sent: 'RGBW => Gain: 255/255, red: 160, green: 160, blue: 160'

continous output is enabled this time! Do you perhaps know what is my fault this time? Did I maybe broke this Lolin also or is it another stupid mistake?

**Reply by awawa-dev** [reply]
Seems so. You didn't stop the LED device after a few seconds of running as it was supposed to be done to give it a chance to get any frame... all this happened around 2023-04-16T15:37:10.73 +/- 10ms. Nevertheless the communication is probably fine.

**Reply by awawa-dev** [reply]
>  another try to use my pcb I damaged the lolin (Its not regognised by the raspi nor my pc again).

the board is not detected even if you trigger bootloader mode using board buttons in the correct order?

**Reply by skalpjaeger** [reply]
nope... unfortunately, tried it several times... But I ordered a new one, hope it arrives soon

**Reply by awawa-dev** [reply]
You need to carefully inspect your electrical system to find out what damaged it. This board doesn't die just like that. This usually happens when +5V is somehow applied to one of its pinouts.

**Reply by skalpjaeger** [reply]
I will! hopefully the new board arrives soon 😄

**Reply by skalpjaeger** [reply]
Finally I found the problem... And it was not the Lolin it seems that I killed my LED Strip :/ I don't know what was going wrong in my last attempt, but with the new Lolin, which arrived today, it was not working either. So I thought maybe its the raspi but also with my Mac the strip was dark. So I soldered Dataline, 5V and ground to the rest of the strip and this was finally working. I really don't know what I did that it killed the lolin and the whole strip... Well at least I can buy the cold blue then oO

**Reply by awawa-dev** [reply]
Most of these LED strips come with cables (+5/data/gnd) and plug already connected, but it is of low quality and can cause a short circuit (e.g. shorting +5V to the data line) when moving. Maybe it happened here? If so you can try to replace the first 2-3 LEDs (cut off and solder a new piece) in the non-working segment, it's quite simple if you have spare.

**Reply by skalpjaeger** [reply]
you are a fucking genius!!! Thanks you a thousand times!!!

**Comment by skalpjaeger** [comment]
Thanks to you, it is working now (Only with one segment to the level shifter connected, the other direct to the esp because if I connect both segments to the level shifter the second segment is flickering. It was working before I mounted the frame to the TV I don't know what I did this time. Perhaps the middle ground loosed connection with the level shifter. Anyway when im motivated I will solder a new pcp but for now im just happy that it works). But my "problem" is that when I activate HDR in HyperHDR settings, my Grabber (MS2130) did not show any video. Only when HDR is not active, it is working as expected... Do you have an explanation for this?

**Reply by awawa-dev** [reply]
Probably broken custom LUT table.
